src/main/java/players/mcts/PISingleTreeNode.java

```diff
@@ -0,0 +1,149 @@
+package players.mcts;
+
+import core.AbstractGameState;
+import core.actions.AbstractAction;
+import players.PlayerConstants;
+
+import java.util.*;
+
+import static players.PlayerConstants.*;
+
+public class PISingleTreeNode extends SingleTreeNode {
+    SingleTreeNode[] roots;
+    AbstractGameState state;
+    int numDeterminizations = 1;
+    MCTSPlayer mctsPlayer;
+    double epsilon = 1e-6;
+    //Action Stats variable to hold aggregated values
+    Map<AbstractAction, ActionStats> accumulatedActionStats;
+
+    //Instantiate A new Tree Node for PI-MCTS
+    public PISingleTreeNode(MCTSPlayer player, AbstractGameState state, Random rnd) {
+        this.decisionPlayer = state.getCurrentPlayer();
+        this.params = player.getParameters();
+        this.forwardModel = player.getForwardModel();
+        this.rnd = rnd;
+        this.state = state;
+        mctsPlayer = player;
+        MASTStatistics = new ArrayList<>();
+        for (int i = 0; i < state.getNPlayers(); i++)
+            MASTStatistics.add(new HashMap<>());
+        if (params.useMASTAsActionHeuristic) {
+            params.actionHeuristic = new MASTActionHeuristic(MASTStatistics, params.MASTActionKey, params.MASTDefaultValue);
+        }
+        numDeterminizations = params.numDeterminizations;
+        roots = new SingleTreeNode[params.numDeterminizations];
+
+    }
+
+    @Override
+    public void mctsSearch(long initialisationTime) {
+        initialiseRootMetrics();
+        initialisationTimeTaken = initialisationTime;
+        //Create Determinized Trees and Run MCTS on the Determinized Trees
+        for (int i = 0; i < params.numDeterminizations; i++) {
+            roots[i] = SingleTreeNode.createRootNode(mctsPlayer, state.copy(state.getCurrentPlayer()), rnd, mctsPlayer.getFactory());
+            roots[i].mctsSearch(initialisationTime);
+        }
+    }
+    //Returns the best action after search based on the chosen AggrergationPolicy
```

Comment:
I reckon we can rely on the mctsSearch in the SingleTreeNode for this (and reduce code repetition a little)